### PR TITLE
Refactor FindEntry and use best-fit .ccls

### PR DESCRIPTION
**Pull request by MaskRay**
Fixes an issue with hierarchical .ccls found by @Riatre in #384

### CODE REVIEW

File: src/project.cc [hunk 1]
```diff
@@ -470,94 +470,99 @@ void Project::Load(const std::string &root) {
...
-  ret.is_inferred = true;
-  ret.filename = path;
-  if (!best) {
-    if (ret.root.empty())
-      ret.root = g_config->fallbackFolder;
-    ret.directory = ret.root;
-    ret.args = GetFallback(path);
+  bool append;
+  if (best_dot_ccls_args && !(append = AppendToCDB(*best_dot_ccls_args))) {
+    // Override the compdb entry.
```
Comment (Riatre) [inline src/project.cc]: It is stated on [wiki](https://github.com/MaskRay/ccls/wiki/Project-Setup#compile_commandsjson-1) that 
> By default .ccls compiler flags are applied only to files not listed in compile_commands.json.

However this is going to override the compile db entry.

Comment (MaskRay) [reply src/project.cc]: This rule may be confusing. I think it is probably clearer to let `.ccls` override `compile_commands.json` if the first option is not `%compile_commands.json`

**Comment by Riatre** [comment]
Yeah. Depends on actual use case both new or old behavior may be wanted
(e.g. "use compile_commands.json primarily and .ccls as fallback" vs "use
special -D flag configurations which is meaningless for actual builds in
some directories") so I have no opinion on this, both are okay for my use
case, just not sure if it is intended or not.

Fangrui Song <notifications@github.com> 于2019年4月30日周二 下午7:38写道：

**Comment by MaskRay** [comment]
I've reinstated the old behavior. The user can just delete `compile_commands.json` (likely a symlink) to make `.ccls` win. I end up with more refactoring as it turns out that lazy `clang.extraArgs` appending improves readability.

**Pull request merged**